tests/test_core.py

```diff
@@ -4044,6 +4044,111 @@ def test_dylink_hyper_dupe(self):
       full = run_js('src.cpp.o.js', engine=JS_ENGINES[0], full_output=True, stderr=STDOUT)
       self.assertContained("warning: trying to dynamically load symbol '_sideg' (from 'third%s') that already exists" % dylib_suffix, full)
 
+  @needs_dlfcn
+  def test_dylink_dso_needed(self):
+    # test that linking to shared library B, which is linked to A, loads A as well.
+    # main is also linked to C, which is also linked to A. A is loaded/initialized only once.
+    #
+    #          B
+    #   main <   > A
+    #          C
+    open('liba.cpp', 'w').write(r'''
+      #include <stdio.h>
+      #include <emscripten.h>
+
+      static const char *afunc_prev;
+
+      EMSCRIPTEN_KEEPALIVE void afunc(const char *s);
+      void afunc(const char *s) {
+        printf("a: %s (prev: %s)\n", s, afunc_prev);
+        afunc_prev = s;
+      }
+
+      struct ainit {
+        ainit() {
+          puts("a: loaded");
+        }
+      };
+
+      static ainit _;
+    ''')
+
+    open('libb.cpp', 'w').write(r'''
+      #include <emscripten.h>
+
+      void afunc(const char *s);
+      EMSCRIPTEN_KEEPALIVE void bfunc();
+      void bfunc() {
+        afunc("b");
+      }
+    ''')
+
+    open('libc.cpp', 'w').write(r'''
+      #include <emscripten.h>
+
+      void afunc(const char *s);
+      EMSCRIPTEN_KEEPALIVE void cfunc();
+      void cfunc() {
+        afunc("c");
+      }
+    ''')
+
+    self.set_setting('TOTAL_MEMORY', 64 * 1024 * 1024)  # XXX default 16MB is not enough
```

Comment:
Yes - with the following debug patch:

```diff
diff --git a/src/support.js b/src/support.js
index 0d8efe1b9..01847d119 100644
--- a/src/support.js
+++ b/src/support.js
@@ -136,6 +136,7 @@ function fetchBinary(url) {
 // flags.global and flags.nodelete are handled every time a load request is made.
 // Once a library becomes "global" or "nodelete", it cannot be removed or unloaded.
 function loadDynamicLibrary(lib, flags) {
+  console.error("loadDynamicLibrary", lib);
   // when loadDynamicLibrary did not have flags, libraries were loaded globally & permanently
   if (flags === undefined) {
     flags = {global: true, nodelete: true}
@@ -345,6 +346,8 @@ function loadWebAssemblyModule(binary, flags) {
   var tableSize = getLEB();
   var tableAlign = getLEB();
 
+  console.error("loadWebAssemblyModule, memorySize:", memorySize);
+
   // shared libraries this module needs. We need to load them first, so that
   // current module could resolve its imports. (see tools/shared.py
   // WebAssembly.make_shared_library() for "dylink" section extension format)
diff --git a/tests/runner.py b/tests/runner.py
index ec25ae364..a2ad9d0e5 100755
--- a/tests/runner.py
+++ b/tests/runner.py
@@ -612,6 +612,7 @@ def run_generated_code(self, engine, filename, args=[], check_timeout=True, outp
       os.chdir(cwd)
     out = open(stdout, 'r').read()
     err = open(stderr, 'r').read()
+    print('stderr:\n', err)
     if engine == SPIDERMONKEY_ENGINE and self.get_setting('ASM_JS') == 1:
       err = self.validate_asmjs(err)
     if output_nicerizer:
```

it becomes clear:

```
wasm@deco:~/emsdk/emscripten/incoming$ ./tests/runner.py binaryen1.test_dylink_dso_needed
test_dylink_dso_needed (test_core.binaryen1) ... stderr:
 loadDynamicLibrary libb.wasm
loadWebAssemblyModule, memorySize: 5242896
loadDynamicLibrary liba.wasm
loadWebAssemblyModule, memorySize: 5242928
loadDynamicLibrary libc.wasm
loadWebAssemblyModule, memorySize: 5242896
loadDynamicLibrary liba.wasm

...
```

in other words every library is so compiled that it tells the linker to reserve ~ 5MB of memory for every module. For 3 libraries (A,B,C) this gives ~15MB that should be only reserved. Probably the rest 1MB is not enough to keep everything else.

Reply:
That seems very strange to me.  Maybe I'm missing something but why would these modules need 5Mb of static data?  They have almost no data at all in them right?

Reply:
Right, that's strange, but this is what emscripten currently emits. I can see that `make_shared_library()` extracts this from `var STATIC_BUMP = ...`

https://github.com/kripken/emscripten/blob/a5ef38a9/tools/shared.py#L2958-L2959

but I had not debugged further. For the reference there is a sign that similar problems were there before, e.g. here:

https://github.com/kripken/emscripten/blob/a5ef38a9/tests/test_core.py#L3979

but there it is added without any explanation.

Reply:
/cc @kripken (who added this in a569b65b)

Reply:
Yes, I added

```
# XXX in wasm each lib load currently takes 5MB; default TOTAL_MEMORY=16MB is thus not enough
```